packages/vite/src/node/plugins/asset.ts

```diff
@@ -414,6 +414,7 @@ const shouldInline = (
 ): boolean => {
   const environment = pluginContext.environment
   const { assetsInlineLimit } = environment.config.build
+  if (id.endsWith('?no-inline')) return false
```

Comment:
We should check using something like this so it's more robust: https://github.com/vitejs/vite/blob/3ceb738c2b0fa0208c861b4f5570a81e193974d0/packages/vite/src/node/plugins/css.ts#L205

Also do we need to handle `?inline` here? I don't think we support that to force inline at the moment.

Reply:
I added the regex, though I'm not sure if there's a situation where there would be several query parameters 🤔...

I also added the `?inline` handling, as you mentioned.

I took the liberty to export the `svgToDataURL` function to a proposer `svg.ts` file, as the `asset.ts` file was already quite big, and it was a bit too specific for this file. Please tell me if I should revert this change.

Reply:
> though I'm not sure if there's a situation where there would be several query parameters

You can[ import assets as url](https://vite.dev/guide/features.html#static-assets) with the `?url` query, so there could be `?url&no-inline` happening.

> I took the liberty to export the `svgToDataURL` function to a proposer `svg.ts` file, as the `asset.ts` file

I think it's better to have `svgToDataURL` in the `asset.ts` file. We have a lot of monster files that really should be broken up, but it's a bit too late without directory restructuring now, so it's better to keep it consistent and keep it in a single file for now.

> lso, tell me if you'd like to see an updated documentation with the `?inline` (it's already covered somewhere else though). Or a test with the `?inline`, like with a svg fragment I guess?

Documentation and test for `?inline` would be great! I think we have several `?inline` support too, but those are for CSS and workers, which isn't exactly the same as assets. It might also be good to test with PNGs instead of SVGs perhaps. We'd like to change the behaviour of SVG inlining in dev (https://github.com/vitejs/vite/issues/18034#issuecomment-2439258851), so that'll make for less conflict.

Reply:
> You can[ import assets as url](https://vite.dev/guide/features.html#static-assets) with the ?url query, so there could be ?url&no-inline happening.

Thanks, I understand!

> but it's a bit too late without directory restructuring now, so it's better to keep it consistent and keep it in a single file for now.

I disagree with this one, but you be the juge... I reverted the change 🙂

> It might also be good to test with PNGs instead of SVGs perhaps

I added the PNG test for ?inline since it makes more sense. However, the ?no-inline intent is for SVG, so I let it on the test. Please tell me if I should put a PNG instead (however, PNG are not inlined by default, so it doesn't test much...)

Note: I also add the `?inline` check for dev server. Without it, there will be a difference between `vite dev` and `vite build`. I think if you force inlining, you want to see the result during dev, not only during build.

So I had to add an await, I'm not 100% sure if this can have a performance impact 🤔...

Reply:
> I added the PNG test for ?inline since it makes more sense. However, the ?no-inline intent is for SVG, so I let it on the test. Please tell me if I should put a PNG instead (however, PNG are not inlined by default, so it doesn't test much...)

That's fine 👍 

> Note: I also add the `?inline` check for dev server. Without it, there will be a difference between `vite dev` and `vite build`. I think if you force inlining, you want to see the result during dev, not only during build.

That also makes sense. We can't really avoid the readFile there, and we need to to handle SVG inlining in dev in the future anyways.